### DISCUSSION TITLE
How to type snapshots?

### DISCUSSION THREAD

**Discussion by skokenes** [pmndrs/valtio#451] [Q&A] [answered]
Hi, I am using valtio with TypeScript. I set up a proxy state with some nested type, like 
```javascript
const state = proxy<{ a: TypeA }>({ a: {} });
```
When I consume the state as a snapshot using `useSnapshot` in my components, I want to pass `snapshot.a` to functions that are expecting the type `TypeA`. However, I am getting TS errors because `snapshot.a` is now typed as a `readonly` variant of my `TypeA`.

Is there some sort of utility type I can use to resolve this issue? I tried modifying my functions to expect a type like `ReadOnly<TypeA>`, but that still did not match `snapshot.a`'s type.

**Comment by dai-shi** [comment, marked as answer]
A snapshot is readonly, and if you want to loosen (lie) the type, please assert it.
```js
const snap = useSnapshot(state) as typeof state
```

This was controversial. Please see #327 for more context.

**Reply by lmaero** [reply]
Hi @dai-shi , thanks for this amazing state management library.

Is this still considered the best way or can you expose the Snapshot<T> as a type?

**Reply by dai-shi** [reply]
Yes, we will expose it in v2. Please try v2-beta #810 (#856). `npm install valtio@next`

But, `as typeof state` is still considered the best way if you want to loosen the type.

**Reply by lmaero** [reply]
Just did the change, it feels more intuitive to type things as Snapshot<T>; excellent, thank you. 

By any chance do you have any docs about -beta version? Just discovered this library, learning a lot!

**Reply by dai-shi** [reply]
We have a small migration guide in the `v5` branch.

**Reply by jaens** [reply]
`as typeof state` does not work except in trivial cases, so I still think another solution is required.

eg.:
```typescript
interface State {
  errors: string[];
}

const state = proxy<State>({errors: []});
// Fails with type error due to not allowing casts between `readonly string[]` and `string[]`
const snapshot = useSnapshot(state) as typeof state;
```

(this is with `tsconfig.json` » `"strict": true`)

The alternative `useSnapshot(state) as unknown as typeof state;` works, but since it's basically not safe at all, it's best left inside a wrapper function.


The wrapper function does not need the `unknown` cast though, since TypeScript strict cast checking does not work for generics...

```typescript
export function useSnapshot_cast<T extends object>(state: T): T {
    return useSnapshot(state) as T;
}
```

**Reply by dai-shi** [reply]
`as unknown as` seems a good workaround. removing `readonly` is not very safe in the first place.

btw, I'm curious why it requires `unknown`.
This seems to work: 

```ts
type X =  { readonly arr: readonly number[] }

const x: X = { arr: [] }
const y = x as { arr: number[] }
```
https://tsplay.dev/mLdA2m

**Reply by jaens** [reply]
Right, it does actually work... :facepalm:  Unfortunately I simplified the example for the discussion, it was originally more complicated, should've saved the exact configuration that caused it.
